### PR TITLE
[Feat] Pinned tooltip + Compare

### PR DESCRIPTION
**Pull request by b2kdaman**
Related to: https://github.com/keplergl/kepler.gl/issues/1083

### CODE REVIEW

**Comment by b2kdaman** [comment]
image

File: src/constants/tooltip.js [hunk 1]
```diff
@@ -169,3 +169,8 @@ export const TOOLTIP_FORMATS = {
     type: TOOLTIP_FORMAT_TYPES.DATE_TIME
   }
 };
+
+export const COMPARE_TYPES = {
+  DECIMAL: 'decimal',
```
Comment (heshan0131) [inline src/constants/tooltip.js:174]: shouldn't this be `absolute` vs `percentage`

File: src/components/map/layer-hover-info.js [hunk 2]
```diff
@@ -81,7 +113,27 @@ const EntryInfoRow = ({item, fields, data}) => {
...
     : parseFieldValue(data[valueIdx], field.type);
-  return <Row name={item.name} value={displayValue} />;
+  let displayDeltaValue = null;
+  if (
+    primaryData &&
+    (field.type === ALL_FIELD_TYPES.integer || field.type === ALL_FIELD_TYPES.real)
+  ) {
+    const deltaValue =
+      compareType === COMPARE_TYPES.PERCENTAGE
+        ? 1 - data[valueIdx] / primaryData[valueIdx]
```
Comment (heshan0131) [inline src/components/map/layer-hover-info.js:123]: is this the right math?  shouldn't it be ` data[valueIdx] / primaryData[valueIdx] - 1`?

File: src/components/map/layer-hover-info.js [hunk 3]
```diff
@@ -81,7 +113,27 @@ const EntryInfoRow = ({item, fields, data}) => {
...
+        : data[valueIdx] - primaryData[valueIdx];
+    const deltaFormat =
+      compareType === COMPARE_TYPES.PERCENTAGE
+        ? TOOLTIP_FORMATS.DECIMAL_PERCENT_FULL_2[TOOLTIP_KEY]
+        : TOOLTIP_FORMATS.DECIMAL_DECIMAL_FIXED_3[TOOLTIP_KEY];
+    displayDeltaValue = getFormatter(item.format || deltaFormat)(deltaValue);
+    const deltaFirstChar = displayDeltaValue.toString().charAt(0);
```
Comment (heshan0131) [inline src/components/map/layer-hover-info.js:130]: why not add another format type: `TOOLTIP_FORMATS.DELTA` and use d3.format('+')(value)

File: src/components/side-panel/interaction-panel/tooltip-config.js [hunk 4]
```diff
@@ -107,12 +132,52 @@ function TooltipConfigFactory(DatasetTag) {
...
+              }
+              selectedItems={config.compareType || COMPARE_TYPES.DECIMAL}
+              options={Object.values(COMPARE_TYPES)}
+              multiSelect={false}
+              searchable={false}
+              inputTheme={'secondary'}
+              onChange={option => {
+                const newConfig = {
+                  ...config,
+                  compareType: translatedCompareTypes[option]
```
Comment (heshan0131) [inline src/components/side-panel/interaction-panel/tooltip-config.js:169]: savein the translated `compareType` to config seems like a bad idea

File: src/constants/tooltip.js (hunk 1 above)
Comment (b2kdaman) [reply src/constants/tooltip.js:174]: Yes! Lost in translation I guess.

File: src/components/side-panel/interaction-panel/tooltip-config.js [hunk 5]
```diff
@@ -107,12 +132,52 @@ function TooltipConfigFactory(DatasetTag) {
...
+              displayOption={d =>
+                intl.formatMessage({
+                  id: `compare.types.${d}`
+                })
+              }
+              selectedItems={config.compareType || COMPARE_TYPES.DECIMAL}
+              options={Object.values(COMPARE_TYPES)}
+              multiSelect={false}
+              searchable={false}
+              inputTheme={'secondary'}
```
Comment (heshan0131) [inline src/components/side-panel/interaction-panel/tooltip-config.js:165]: you can use `getOptionValue={d => d}` to have the `onChange` function return the untranslated value. hence no need to do things as below

File: src/components/side-panel/interaction-panel/tooltip-config.js (hunk 4 above)
Comment (heshan0131) [inline src/components/side-panel/interaction-panel/tooltip-config.js:169]: add `comparType` to initial tooltipConfig, and type definition

Also `comparType` is not added to schema, it won't be saved

File: src/components/map/layer-hover-info.js (hunk 2 above)
Comment (b2kdaman) [reply src/components/map/layer-hover-info.js:123]: Welp I guess I'm that bad at math haha

File: src/components/side-panel/interaction-panel/tooltip-config.js (hunk 4 above)
Comment (b2kdaman) [reply src/components/side-panel/interaction-panel/tooltip-config.js:169]: It's actually default value. Not translated.  You did the same with layer blending ^^

File: src/components/map/layer-hover-info.js (hunk 3 above)
Comment (b2kdaman) [reply src/components/map/layer-hover-info.js:130]: If we have % in the string it would be NaN

File: src/components/map-container.js [hunk 6]
```diff
@@ -249,54 +276,64 @@ export default function MapContainerFactory(MapPopover, MapControl, Editor) {
...
-            layerHoverProp={layerHoverProp}
-            coordinate={
-              interactionConfig.coordinate.enabled && ((pinned || {}).coordinate || coordinate)
-            }
-            freezed={Boolean(clicked || pinned)}
-            onClose={this._onCloseMapPopover}
-            mapW={mapState.width}
-            mapH={mapState.height}
-          />
+          {(clicked || pinned) && (
```
Comment (heshan0131) [inline src/components/map-container.js:315]: the logic here has become way too complicated clarify with varible name will be helpful
what is `clicked || pinned` and what is `(compareMode || (!clicked && !pinned))` can it be named to `hasTooltip` `hasComparisonTooltip`?

Comment (b2kdaman) [reply src/components/map-container.js:315]: Got it

**Comment by b2kdaman** [comment]
@heshan0131 I 
Screen Recording 2020-06-15 at 12 56 37 PM
added this

**Comment by heshan0131** [comment]
@b2kdaman I moved some component logic into utils and added tests based in this commit https://github.com/keplergl/kepler.gl/commit/70df9f4bd01c75eabaef7c879e33449b92caa968

Please cherrypick it onto your branch, then we are good to land

**Pull request merged**